net: Add unit testing of node eviction logic

practicalswift opened the pull request:
Add unit testing of node eviction logic.

Closes #19966.

practicalswift committed:
net: Move eviction node selection logic to SelectNodeToEvict(...)

practicalswift committed:
test: Add unit testing of node eviction logic

jonatack commented:
Thanks for working on adding testing here. I'll try to review this soon.

laanwj commented on src/net.cpp:
```diff
@@ -1016,7 +967,51 @@ bool CConnman::AttemptToEvictConnection()
+    return vEvictionCandidates.front();
+}
+
+/** Try to find a connection to evict when the node is full.
```
Please move this doc-comment to the header, it's an exported function now.

laanwj commented:
Concept ACK.

I sometimes wonder if it would make sense to move "semi-internal" things that are only exposed externally for unit testing, like in this case "NodeEvictionCandidate" and "SelectNodeToEvict", to a separate set of headers. But not here anyhow.

dhruv commented on src/net.cpp:
```diff
@@ -1016,7 +967,51 @@ bool CConnman::AttemptToEvictConnection()
+                                               HasAllDesirableServiceFlags(node->nServices),
+                                               peer_relay_txes, peer_filter_not_null, node->nKeyedNetGroup,
+                                               node->m_prefer_evict, node->addr.IsLocal()};
+            vEvictionCandidates.push_back(candidate);
```
Would it make sense to use `emplace_back` here?

practicalswift replied:
I'm afraid that won't be possible until C++20¹ :) See [P0960: "Allow initializing aggregates from a parenthesized list of values" ](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0960r0.html).

¹ Assuming you mean without adding a new ctor to `NodeEvictionCandidate`.

dhruv reviewed (commented):
utACK `2c62fb6`. Thank you, @practicalswift  for making the code more testable. Also for exposure to idiomatic modern C++. I will checkout the branch and run soon.

jnewbery commented on src/net.cpp:
```diff
@@ -910,43 +896,8 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
- *   simultaneously better at all of them than honest peers.
- */
-bool CConnman::AttemptToEvictConnection()
+NODISCARD Optional<NodeEvictionCandidate> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates)
```
This is a silent merge conflict with master. `NODISCARD` should be replaced with `[[nodiscard]]`

jnewbery replied:
Why return a `NodeEvictionCandidate` optional here, when only the NodeId is used? I think it'd be better to return an `Optional<NodeId>`

jnewbery commented on src/net.cpp:
```diff
@@ -1016,7 +967,51 @@ bool CConnman::AttemptToEvictConnection()
+                                               node->m_prefer_evict, node->addr.IsLocal()};
+            vEvictionCandidates.push_back(candidate);
+        }
+        node_to_evict = SelectNodeToEvict(std::move(vEvictionCandidates));
```
Why is this called with `cs_vNodes` held? Previously, we'd release the lock before running through this logic.

practicalswift replied:
Good point. It can be moved to the outer scope. Thanks!

jnewbery commented on src/net.cpp:
```diff
@@ -1016,7 +967,51 @@ bool CConnman::AttemptToEvictConnection()
+                                               node->m_prefer_evict, node->addr.IsLocal()};
+            vEvictionCandidates.push_back(candidate);
+        }
+        node_to_evict = SelectNodeToEvict(std::move(vEvictionCandidates));
```
What's the reason for the `std::move` here? Why not just pass this as an lvalue reference?

jnewbery replied:
I understand why you'd want to pass by ref, but there's no benefit to passing by rvalue. This code suggests to me that `SelectNodeToEvict()` is going to keep ownership of `vEvictionCandidates` and you're using move semantics to avoid a copy, but that's not actually happening here.

There's no harm in using `&&` and `move` here, but it's unnecessary and possibly confusing.

jnewbery commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,129 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+{
+    std::vector<NodeEvictionCandidate> candidates;
+    for (int i = 0; i < n_candidates; ++i) {
+        candidates.push_back({/* id */ static_cast<NodeId>(random_context.randrange(100)),
```
This can create duplicate node ids, which isn't possible in the product code.

jnewbery commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,129 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+    if (!evicted_node) {
+        return false;
+    }
+    return std::find(node_ids.begin(), node_ids.end(), evicted_node->id) != node_ids.end();
```
Does this break if there are multiple eviction candidates with the same id?

jnewbery commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,129 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+// the current eviction logic will find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+const int GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE = 29;
```
```suggestion
constexpr int GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE{29};
```

jnewbery commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,129 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+            // protected from eviction.
+            BOOST_CHECK(!IsEvicted(
+                number_of_nodes, [number_of_nodes](NodeEvictionCandidate& candidate, const size_t idx) {
+                    candidate.id = idx;
```
I think if you just constructed the `NodeEvictionCandidate`s with sequential node id numbers, you wouldn't need to do this here.

practicalswift commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,129 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+// the current eviction logic won't find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+const int GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW = 20;
```
Yes, but I'm afraid I cannot help you with that at the moment: so far it is a purely an empirical result :D

Counterexamples or proofs welcome! :)

dhruv replied:
Possible reason: the protection at [net.cpp::961](https://github.com/bitcoin/bitcoin/blob/fabecce/src/net.cpp#L961) for up to 8 non-tx-relay peers may, or may not apply to the randomly generated eviction candidates since `!n.fRelayTxes && n.fRelevantServices` will evaluate to randbool. So we cannot count them in the _guaranteed_ eviction?

practicalswift commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,129 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+// the current eviction logic will find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+const int GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE = 29;
```
I think that is a good idea but I kind of like that this PR doesn't touch the actual selection code (keeping the diff at a minimum). I'd like to keep it that way but I'd be glad to review a change like the one suggested in a trivial "introduce constants" PR:).

MarcoFalke commented:
tsan failue can be ignored or fixed by a rebase

practicalswift commented:
@MarcoFalke Thanks! Rebased!

jnewbery commented on src/net.cpp:
```diff
@@ -9,6 +9,7 @@
 
 #include <net.h>
 
+#include <attributes.h>
```
No longer needed since you removed `NODISCARD`

practicalswift replied:
Fixed!

jnewbery commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,125 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+    return candidates;
+}
+
+bool IsEvicted(const std::vector<NodeEvictionCandidate>& candidates, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
```
I think this function and the other `IsEvicted()` could use a very short comment to explain the interface (here: returns true if any of the node ids in `node_ids` are selected for eviction).

practicalswift replied:
Fixed!

jnewbery reviewed (commented):
Very strong concept ACK. This area of the code is definitely undertested. The structure of the unit tests (passing in a setup lambda function to set the relevant parameters in the eviction candidates) is really nice.

MarcoFalke commented:
needs rebase

practicalswift commented:
> I understand why you'd want to pass by ref, but there's no benefit to passing by rvalue. This code suggests to me that SelectNodeToEvict() is going to keep ownership of vEvictionCandidates and you're using move semantics to avoid a copy, but that's not actually happening here.

I've now reverted to the original version (alternative 1 below), since [your suggestion](https://github.com/bitcoin/bitcoin/pull/20477#discussion_r537411893) and [sipa's suggestion](https://github.com/bitcoin/bitcoin/pull/19972#discussion_r517734066) are mutually exclusive :)

I'll let others chime in regarding the choice between the three alternatives:

```
Alternative 1. Optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>)
Alternative 2. Optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&)
Alternative 3. Optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&&)
```

Trying to summarize the pros/cons:
* Alternative 1. Simple and no gotchas. Does an extraneous copy.
* Alternative 2. No extraneous copy. Possibly surprising that a function named `SelectNodeToEvict` modifies its input.
* Alternative 3. No extraneous copy. Risk of accidental use-after-move in the future. Possibly confusing (see https://github.com/bitcoin/bitcoin/pull/20477#discussion_r537411893).


Perhaps the members of the review club could chime in. Some questions for the review club: Do you agree with the pros/cons of each alternative? Which alternative do you find easier to review? Which alternative do you prefer? Is it possible to quantify the real-world impact of the extraneous copy in a theoretical worst-case scenario?

jnewbery commented:
> since your suggestion and sipa's suggestion are mutually exclusive :)

I don't think they are mutually exclusive. There are two questions here:

1. Should the argument be passed by value or by reference?
2. If passed by reference, should it be passed by lvalue reference or rvalue reference?

I don't think there's any disagreement about (1). We should pass by reference to avoid a copy.

For (2), I think it's slightly odd to pass by rvalue reference when it's not necessary to do so, but that's more of a stylistic thing. I see now that you're calling the function with a prvalue in the tests, which means you can't pass by lvalue reference. My preference would therefore be to revert to what you had before.

jnewbery commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,129 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+}
+
+// Returns true if any of the node ids in node_ids are selected for eviction.
+bool IsEvicted(const std::vector<NodeEvictionCandidate>& candidates, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> shuffled_candidates = candidates;
+    Shuffle(shuffled_candidates.begin(), shuffled_candidates.end(), random_context);
+    const Optional<NodeId> evicted_node_id = SelectNodeToEvict(shuffled_candidates);
```
Any reason not to shuffle in place?

```suggestion
bool IsEvicted(std::vector<NodeEvictionCandidate>& candidates, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
{
    Shuffle(candidates.begin(), candidates.end(), random_context);
    const Optional<NodeId> evicted_node_id = SelectNodeToEvict(candidates);
```

practicalswift commented:
@jnewbery 

Updated accordingly. 

Personally I prefer sipa's suggested alternative since it makes it clear from reading the code at the call site that `vEvictionCandidates` is not intended to be used after the `SelectNodeToEvict` call.

Hopefully any use-after-move would be quite obvious.

jnewbery commented:
> Personally I prefer the std::vector<NodeEvictionCandidate>&& alternative in this case since it makes it clear from reading the code at the call site that vEvictionCandidates is not intended to be used after the SelectNodeToEvict call.

Fine by me!

jnewbery commented:
utACK cadd93f35d1bbf662e348a0dee172cdf4af6a903

narula commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,128 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+// the current eviction logic will find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+constexpr int GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE{29};
```
Please add some justification for these numbers, even if it's a description of the simulation/test you ran to get them. I don't understand how this 29 is guaranteed and the 20 is guaranteed below. 

This reply makes it sound like it's *not* guaranteed?

https://github.com/bitcoin/bitcoin/pull/20477#discussion_r536421598

narula replied:
Good discussion in PR review club about it today and @jonatack pointed out justification for the 20 is in `test/functional/p2p_eviction.py`.

narula commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,128 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+            if (number_of_nodes >= GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE) {
+                BOOST_CHECK(SelectNodeToEvict(GetRandomNodeEvictionCandidates(number_of_nodes, random_context)));
+            } else {
+                continue;
```
I think you're not running the following checks if the number of nodes is between the min and the max because of this `continue`. Is that the intention?

MarcoFalke replied:
I don't get this either

troygiorshev replied:
I've proposed an alternative that might be easier to parse.
https://github.com/bitcoin/bitcoin/pull/20477/files#r539574518

troygiorshev replied:
It's "don't run the checks if the number of nodes is less than the max".  But, I'm not sure why we would do this...

practicalswift replied:
@narula Oh, good catch! The `continue` should be dropped. I don't remember what the apparently incorrect reasoning behind it was TBH :)

Thanks for catching it!

maflcko reviewed (commented):
Concept ACK

troygiorshev commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,128 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+
+    for (int i = 0; i < NODE_EVICTION_TEST_ROUNDS; ++i) {
+        for (int number_of_nodes = 0; number_of_nodes < NODE_EVICTION_TEST_UP_TO_N_NODES; ++number_of_nodes) {
+            // Verify correctness of GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE
+            // and GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW.
+            if (number_of_nodes <= GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW) {
+                BOOST_CHECK(!SelectNodeToEvict(GetRandomNodeEvictionCandidates(number_of_nodes, random_context)));
+            }
+            if (number_of_nodes >= GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE) {
+                BOOST_CHECK(SelectNodeToEvict(GetRandomNodeEvictionCandidates(number_of_nodes, random_context)));
+            } else {
+                continue;
+            }
```
I'm a HUGE supporter of the early return pattern (AKA bouncer pattern) that you're using here.  However, maybe this alternative is a little easier to parse.

```suggestion
            if (number_of_nodes < GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE) {
            	// Verify correctness of GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW
            	if (number_of_nodes <= GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW) {
            		// We're guaranteed to never evict any nodes
                	BOOST_CHECK(!SelectNodeToEvict(GetRandomNodeEvictionCandidates(number_of_nodes, random_context)));
                }
            	// There is no need to run the sub-tests for fewer than the number of nodes guaranteed for eviction
            	// ^ FWIW I don't agree with this, see below
            	continue;
            }
            // number_of_nodes >= GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE
            // Verify correctness of GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE
            BOOST_CHECK(SelectNodeToEvict(GetRandomNodeEvictionCandidates(number_of_nodes, random_context)));  
```

Honestly though, I think it's just the similarity between the names of `GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE` and `GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW` that is making this so difficult.

All of this said, why have the early return at all?  All of the tests below check for non-eviction.  Can't we run them regardless of the number of nodes?

(Feel free to reword my code comments)

practicalswift commented:
Thanks for reviewing!

Pushed an updated version:

Now skipping the `GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE` and `GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW` checking which isn't particularly important compared to the other tests.

Also fixed an incorrect comment.

Should hopefully be ready for final review :)

jnewbery commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,102 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+                },
+                {0, 1, 2, 3}, random_context));
+
+            // Four non-tx-relay peers that most recently sent us novel blocks
```
The comment here says 'four', but the code below is asserting that 8 non-tx-relay candidates are protected.

practicalswift replied:
Good catch. There are two `nLastBlockTime` based protections: that's why more than four are protected in practice in this specific test case. Now testing the two different `nLastBlockTime` protections a.) independently and b.) jointly.

MarcoFalke commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,102 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+bool IsEvicted(const int number_of_nodes, std::function<void(NodeEvictionCandidate&)> candidate_setup_fn, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates = GetRandomNodeEvictionCandidates(number_of_nodes, random_context);
+    for (size_t i = 0; i < candidates.size(); ++i) {
```
nit: could use ranged for loop?

practicalswift replied:
Oh, of course. A prior version used the index. Thanks!

MarcoFalke commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,102 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+    if (!evicted_node_id) {
+        return false;
+    }
+    return std::find(node_ids.begin(), node_ids.end(), *evicted_node_id) != node_ids.end();
```
how is this different from a less verbose 
```suggestion
    return node_ids.find(*evicted_node_id) != node_ids.end();
```

practicalswift replied:
`std::vector` doesn't have a `find` function. Perhaps you were thinking `std::map` or `std::set`? :)

MarcoFalke commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,102 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+                              /* fBloomFilter */ random_context.randbool(),
+                              /* nKeyedNetGroup */ random_context.randrange(100),
+                              /* prefer_evict */ random_context.randbool(),
+                              /* m_is_local */ random_context.randbool()});
```
style-nit: Adding a trailing comma here would not only make future diffs smaller if they add a new member, but also eat the extraneous whitespace at the beginning of the line.

practicalswift replied:
Good point! Thanks!

maflcko reviewed (approved):
ACK.

Nice test.

practicalswift commented:
Updated.

Addressed feedback from @MarcoFalke and @jnewbery.

Also added a combined test which tests the protections based on `nKeyedNetGroup` (4 peers protected), `nMinPingUsecTime` (8 peers protected), `nLastTXTime` (4 peers protected) and `nLastBlockTime` (8 peers protected) jointly.

Kindly re-review thoroughly :)

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,133 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+        for (int number_of_nodes = 0; number_of_nodes < NODE_EVICTION_TEST_UP_TO_N_NODES; ++number_of_nodes) {
+            // Four nodes with the highest keyed netgroup values should be
+            // protected from eviction.
+            BOOST_CHECK(!IsEvicted(
```
In the previous version of the test, we tested:

1. No eviction happens with <= 20 peers
2. An eviction happens with >= 29 peers
3. If an eviction happens, the protected peers are not evicted

In the new version, we are checking that the protected peers are not evicted in any situation (3). The last test case covers for (1). However, the test will not alert if an eviction does not happen when it should (>= 29 peers). Am I missing something?

Personally, I found the GUARANTEED_EVICTION_THRESHOLD and GUARANTEED_NON_EVICTION_THRESHOLD valuable to test and also as documentation.

practicalswift replied:
@dhruv I can re-add it if someone has time to figure out the actual logic behind the constant 29. I haven't had time to back my empirical observation (via simulation) with proper analysis. Help welcome! :)

dhruv replied:
From PR Review club:

**Q: Why is eviction guaranteed if we have at least 29 eviction candidates?**

The code in net.cpp (https://github.com/bitcoin/bitcoin/blob/fabecce/src/net.cpp#L954-L968) protects at most 28 peers from eviction (4 by net group, 8 by lowest ping time, 4 by last time of novel tx, up to 8 non-tx-relay peers by last novel block time, and 4 more peers by last novel block time). So any additional peers are guaranteed to be candidates for eviction.

**Q: Why is non-eviction guaranteed if we have no more than 20 eviction candidates? Is 20 the highest number of nodes that guarantees non-eviction?**

The protection at [net.cpp::961](https://github.com/bitcoin/bitcoin/blob/fabecce/src/net.cpp#L961) for up to 8 non-tx-relay peers may, or may not apply to the randomly generated eviction candidates since `!n.fRelayTxes && n.fRelevantServices` will evaluate to `randbool`. So we cannot count them in the _guaranteed_ non-eviction? 

So, guaranteed non-eviction is only on 4(CompareNetGroupKeyed) + 8(ReverseCompareNodeMinPingTime) + 4(CompareNodeTXTime) + 4(CompareNodeBlockTime) = 20.

jonatack replied:
@practicalswift the review club meeting log about your PR is here, if useful: https://bitcoincore.reviews/20477

practicalswift replied:
Thanks! Now assert eviction at >= 29 candidates and non-eviction at <= 20 candidates. Please re-review :)

jnewbery replied:
I don't think these <=20/>=29 tests are very useful, will need to be updated whenever any changes are made to the eviction logic, and aren't very realistic (in general, nodes have many more than 29 inbound slots).

practicalswift replied:
I'll happily adjust to the consensus opinion, but I'll let others chime in before changing anything in order to avoid another round of change+revert :)

dhruv replied:
Even if we remove the <=20 assertion, we are already checking non-eviction <=20 (it's implicit in the tests). So the tests will need need to updated when changes are made anyway. I am not sure I understand why that's undesirable.

I think it is useful to test that eviction does indeed happen (not just that protections are provided) when it should. After all, `SelectNodeToEvict` should select a node to evict :) Without the >=29 assertion, does this test guarantee that it does?

All that said, @jnewbery is infinitely more qualified than I am. I'm happy with whatever you all decide and will learn from it.

jnewbery commented:
ACK 6767d63d02cc8a670f3d409db1c78a77fdcb5363

jonatack commented:
Will review properly as soon as my laptop is done with the gitian builds for 3 RCs.

practicalswift committed:
Assert eviction at >= 29 candidates. Assert non-eviction at <= 20 candidates.

practicalswift commented:
Updated to address @dhruv's feedback. Should hopefully be ready for final review. Please re-review :)

dhruv commented:
ACK `1c9b235`

MarcoFalke commented:
cr ACK 6767d63d02cc8a670f3d409db1c78a77fdcb5363 🤠

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

cr ACK 6767d63d02cc8a670f3d409db1c78a77fdcb5363 🤠
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgJLwwAljffgO2Xw1Iez563pRA0ztSBIkuzEBPGf9nYifcbQJsfPMCdLzuu3MPQ
ZgfWPQNmtEuSTNq0R4Hl+XaPmBcoS5LDOTfxJWsQtb+kE0uXVfN1nRu4+d91/Sor
+NT9Ach7YRVvJPOZZc0euxyka0pxLwX3+LdXa1k26gL2teT5knaROnCfSxenrFrO
YIzbBfkn2lm/TMu4FhC2hW+OZon4zfHHCPzJYkQR2FByixLDqIIEOJiOY9ycLJQm
ys9xGzn0lZz0/hw4ApEe2b+lLREn/6YMjPNefo0hWcADAYDbjsvEYBsFm0Z6m5Mw
BXA4lhUPWes5QboZ1LCxrEdxGG10uj8LwIBwESp/KVpZ53w0f9yU+mBQ68bruAHh
5hFAJmriES6q8Ml/q2b3IUq8tXwOBM+2poGR8MtTYV6Obk/3eus2bA33EVsfNzm3
ZXzo/F1yQvXlD1kSDU3Aez1yodeLK3GQNoTKaPIGeIwOM3sPhDhJ6REMYskLM4tt
PxqHvOZz
=Gf1H
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `23398d20d71a2405d911d0dc99e4fc62b592150f49fa6c17afd6d588c418522c  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010823398d20d71a2405d911d0dc99e4fc62b592150f49fa6c17afd6d588c418522cf0107fd7398bc90e0be7a605322137d2ed9108fff01002bd7f23a0a9a6df040ad1fb6a53ce5e08f1045fd9dbbbf008a9f142f6512d6a5e0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0102cc72ffccbaac7120b1f3f99933682b808f1045fd9dbbaf0080e361c96f960f3860083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010461c5cf5d05a38312544cb8d5b6c4b8408f1045fd9dbbaf0087ee78d84418393070083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df0108fcf32ebf4bd8336aeba0cb60420e84d08f020e09db3e429c6645c18d52d92c85db443d9945c22630cb987573b2856fd9aa9aa08f0200b5a8780a62f3f7df85192d5e0105257ef184ac37603bce482a379453152e1e908f1045fd9dbbaf008277420a4c865ad980083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
cr ACK 1c9b235c330fe7b9b71f687ef14cb10fe588172b 👲

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

cr ACK 1c9b235c330fe7b9b71f687ef14cb10fe588172b 👲
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjQWgv/TnEWYjD/Ws2y2WL23agZIEyXZwlxbLCJTWkydKzBnOrPdP1QOBl3DZ95
1ICN5MfBErZtO825TG8NU2VfCesJRIJKV4996nw7GqpN+KfBlkYHNipYfoF9xdHB
J0MT/97/wsrsWGhYEdwx3ftk0CHMNNDMN++N4BSx3h314jGnuNzV4DiUoz2ea9X3
1IAsitmrqa37xIlLIvdLSwRVDcWQyJ7oQ+NUo2KsSztXJ6XEpZcUAqhfRwuS9Ku4
R3DIJHLXKk8PUEXlnmDWVtfnhuhGW4g3V6ZJdUId7OfUr85H/rGSxSeejqSqxLt/
8hGiRsuftelXPfd7SOAjRW5h6x6NFCJBgcVVOYKzjvjyknyH3it3IrPi9WFwLLWS
98fNiUtiE2SsXOUFXxbbyalCZlQols/pBklibFoPNPhfHCT7ZPvazINGBJwkwxVS
GL1IIAKftgsUJhmiSDy/LHfHwXC2vOL2ipdPLJuR7CPLPj9iBZ2picP4h3zRpSo2
lubdxdSX
=z6dl
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `3a47dabfa2f12ebc4fbd136dcd7771912bfcd7ed7ce86ba72f473c364fba70a0  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401083a47dabfa2f12ebc4fbd136dcd7771912bfcd7ed7ce86ba72f473c364fba70a0f010b438fe0e601d86841026faaaaf5a7f5e08fff01071cd1dfe154052ccbeee0f24f9e9acd108f1045fd9dc30f008a0ca2a256c0e69180083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0108866883bd2ed9744c592c7cb9dfcc2e408f1045fd9dc2ff008a95c1c9a664641660083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0109dd05ccbc7815398d0399c92373b228f08f1045fd9dc32f0080f1b2b81731ae62a0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010d9c46dfac07e0e8de83409f86c44694708f1045fd9dc2ff0086ec4bf813030481b0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

MarcoFalke commented:
I ACKed both versions. Let me know which one to merge.

jnewbery commented:
code review ACK 1c9b235c33

Either are fine to merge.

jonatack commented:
wait for meeee 😀

jonatack commented on src/net.cpp:
```diff
@@ -910,43 +895,8 @@ static void EraseLastKElements(std::vector<T> &elements, Comparator comparator,
- *   simultaneously better at all of them than honest peers.
- */
-bool CConnman::AttemptToEvictConnection()
+[[nodiscard]] Optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates)
```
fbba7d8aada5b1d7a63ad4133 add `#include <optional.h>` header

practicalswift replied:
Now including `optional.h`.

Regarding the naming: I see your point but given `NodeEvictionCandidate` inputs I think `SelectNodeToEvict` works.

jonatack commented on src/net.cpp:
```diff
@@ -1016,10 +966,52 @@ bool CConnman::AttemptToEvictConnection()
+            vEvictionCandidates.push_back(candidate);
+        }
+    }
+    Optional<NodeId> node_id_to_evict = SelectNodeToEvict(std::move(vEvictionCandidates));
```
fbba7d8aada5b1d7a63ad4 could be const

practicalswift replied:
Good catch! Fixed!

jonatack commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,147 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+}
+
+// Returns true if any of the node ids in node_ids are selected for eviction.
+bool IsEvicted(std::vector<NodeEvictionCandidate> candidates, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
```
6767d63d02cc8a670f3d409db1 ref to const?
```suggestion
bool IsEvicted(std::vector<NodeEvictionCandidate> candidates, const std::vector<NodeId>& node_ids, FastRandomContext& random_context)
```

practicalswift replied:
Good catch! Immutate is better than mutable: fixed!

jonatack commented on src/test/net_tests.cpp:
```diff
@@ -771,4 +773,147 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+// Create number_of_nodes random nodes, apply setup function candidate_setup_fn,
+// apply eviction logic and then return true if any of the node ids in node_ids
+// are selected for eviction.
+bool IsEvicted(const int number_of_nodes, std::function<void(NodeEvictionCandidate&)> candidate_setup_fn, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
```
6767d63d02cc8a670f3d409 ```suggestion same
bool IsEvicted(const int number_of_nodes, std::function<void(NodeEvictionCandidate&)> candidate_setup_fn, const std::vector<NodeId>& node_ids, FastRandomContext& random_context)
```

practicalswift replied:
Good catch! Immutate is better than mutable: fixed!

jonatack commented:
ACK 1c9b235c330fe7b9b71f modulo a few suggestions

practicalswift commented:
@jonatack Thanks for reviewing! Feedback addressed: please re-review :)

jonatack commented:
ACK fee88237e03c21bf81f21098e6b89ecfa5327cee

Thanks for adding this nicely done coverage! Looking forward to more.

MarcoFalke commented:
ACK fee88237e03c21bf81f21098e6b89ecfa5327cee 🐼

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK fee88237e03c21bf81f21098e6b89ecfa5327cee 🐼
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiu/Av/RLQ2/lM0mJ7uTlvDebTgk57QGAdIs4rv+t9X7wUizqNhEjFHU2KC0HS0
hU6QjHZelMuy0IvIRrcT33LCZkczPQbEQ4ZMAV5VZhuRNvAisi1PHxueX5U+oT83
UKS899fiy7VN3RD7g7E/sMDNuGAo1hSV7MgYQMOlCG8dSJ3bbuF2d1jNEYqJQ7oa
MkmdtB0SwvCOtZyvcsVDI8yC8UJ5r2ybgE+bYfMK9BzMUyWQriWmOnoVQBrnwb5j
Zr7g/4pnq04078EuaBVP981P+Pc4pj6Eo63iIinabl74aLU3hsHa0bnUCkpihBSg
stooipOqKXQsOS/jI1XyElhclE3EolEE6+0QIXFpx3H7AlkovPdF/LeeezH0dVDB
hSl7VrdKrTxP7kjSBKBMOiPbzBdVSqo6ktaqhBI7YXlDepR7V54/PlDs7QEcM91r
3iEvD/wbMFJQZHzHEbmCINPXBwdOh9MleFgm0dgTMor7djOpaxgbYq/2Y5Bx4/4K
s1ZuLUUS
=f3JL
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `fb125172e7abd1e8e484ae4f527e5e86e05737d1b1e006310dca8dbb38203afc  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108fb125172e7abd1e8e484ae4f527e5e86e05737d1b1e006310dca8dbb38203afcf01021bd9625059be8619f831b67c585e18708fff01047bc36d91c428388a4477ef64b7e946e08f1207a5438daa431ce45f7483dec3c3dbf3b7d8834007f7c23a78aa2a27d6e64087508f1045fd9fde2f008063996b0f7f67d2c0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01061190039bdd1fbfbabee96c824f4a40e08f12071c03c42d2b08d21b3b13974e874eb9a28988e8ad3d8ba93d17e43b9faa3f3b708f1045fd9fde1f0089052e62b0b37905f0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010b3f711330b0e9f7c1217942c198ef08108f02028c6ec707950e038eefb921e4bb2ab74cd13c60d03bcc17dbc9ab8cab9c5f20a08f120d9e9f61eeeba3a6da1845028e41b8e7072429c02b168ba997847f36474941b5208f020c0e631cbe0b411022c8fb6a38cf34212b9bbd91125314d7ee3627dd30e03606c08f1045fd9fde2f008e67a6537946812580083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010fea4c6daa41c07668282ff82e88a6e4808f1207d4951d5924eeae2fde3701dafb4f4b7d4da87674f902db740b5f9542a90607208f02083062c158565455dad7a807a546fab2433a251138f8665ab19c4fafc960e120108f0206c7957cbeb8becfd6bbf12fb19d778c276c0f53775dca11bb6d1cfa71985adcb08f1045fd9fde2f008784c47c1cbfbe60a0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
